libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -1464,147 +1678,184 @@ bool AC_PosControl::get_fwd_pitch_is_limited() const
-        return POSCONTROL_OVERSPEED_GAIN_U * _vel_desired_neu_cms.z / _vel_max_down_cms;
+    // If desired descent speed exceeds configured max, scale acceleration/jerk proportionally
+    if (_vel_desired_neu_ms.z < _vel_max_down_ms && !is_zero(_vel_max_down_ms)) {
+        return POSCONTROL_OVERSPEED_GAIN_U * _vel_desired_neu_ms.z / _vel_max_down_ms;
     }
-    if (_vel_desired_neu_cms.z > _vel_max_up_cms && !is_zero(_vel_max_up_cms)) {
-        return POSCONTROL_OVERSPEED_GAIN_U * _vel_desired_neu_cms.z / _vel_max_up_cms;
+
+    // If desired climb speed exceeds configured max, scale acceleration/jerk proportionally
+    if (_vel_desired_neu_ms.z > _vel_max_up_ms && !is_zero(_vel_max_up_ms)) {
+        return POSCONTROL_OVERSPEED_GAIN_U * _vel_desired_neu_ms.z / _vel_max_up_ms;
     }
+
+    // Within normal speed limits — use nominal acceleration and jerk
     return 1.0;
 }
 
-/// initialise ekf xy position reset check
+// Initializes tracking of NE EKF position resets.
 void AC_PosControl::init_ekf_NE_reset()
 {
     Vector2f pos_shift;
     _ekf_ne_reset_ms = _ahrs.getLastPosNorthEastReset(pos_shift);
 }
 
-/// handle_ekf_NE_reset - check for ekf position reset and adjust loiter or brake target position
+// Handles NE position reset detection and response (e.g., clearing accumulated errors).
 void AC_PosControl::handle_ekf_NE_reset()
 {
-    // check for position shift
+    // Check for EKF-reported NE position shift since last update
     Vector2f pos_shift;
     uint32_t reset_ms = _ahrs.getLastPosNorthEastReset(pos_shift);
     if (reset_ms != _ekf_ne_reset_ms) {
+        // Reset NE controller to preserve relative position control during Loiter, PosHold, etc.
+        // This ensures controller output remains continuous after EKF realigns the origin.
 
         // ToDo: move EKF steps into the offsets for modes setting absolute position and velocity
         // for this we need some sort of switch to select what type of EKF handling we want to use
 
-        // To zero real position shift during relative position modes like Loiter, PosHold, Guided velocity and accleration control.
-        _pos_target_neu_cm.xy() = _pos_estimate_neu_cm.xy() + _p_pos_ne.get_error().topostype();
-        _pos_desired_neu_cm.xy() = _pos_target_neu_cm.xy() - _pos_offset_neu_cm.xy();
-        _vel_target_neu_cms.xy() = _vel_estimate_neu_cms.xy() + _pid_vel_ne.get_error();
-        _vel_desired_neu_cms.xy() = _vel_target_neu_cms.xy() - _vel_offset_neu_cms.xy();
+        // Reconstruct target and desired positions relative to new EKF origin
+        _pos_target_neu_m.xy() = _pos_estimate_neu_m.xy() + _p_pos_ne.get_error().topostype();
+        _pos_desired_neu_m.xy() = _pos_target_neu_m.xy() - _pos_offset_neu_m.xy();
+
+        // Reconstruct velocity targets based on PID error and velocity estimate
+        _vel_target_neu_ms.xy() = _vel_estimate_neu_ms.xy() + _pid_vel_ne.get_error() * 0.01;
```

Comment:
it's really tricky that the velocity error is in cm/s.  It seems like this is the only place that we're converting it back to m?

Reply:
Yep, most only care about it's direction.